framework/src/tinyalsa/tinyalsa.c

```diff
@@ -531,49 +535,99 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
 
 	nbytes = pcm_frames_to_bytes(pcm, frame_count);
 
-#ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
-	if (nbytes < pcm->config.period_size) {
-		return -EINVAL;
-	}
 	bufdesc.numbytes = pcm->config.period_size;
-#else
-	if (nbytes < CONFIG_AUDIO_BUFFER_NUMBYTES) {
-		return -EINVAL;
-	}
-	bufdesc.numbytes = CONFIG_AUDIO_BUFFER_NUMBYTES;
-#endif
+
 #ifdef CONFIG_AUDIO_MULTI_SESSION
 	bufdesc.session = pcm->session;
 #endif
 
 	/* If device is not yet started, start now! */
-	if ((!pcm->running) && (pcm_start(pcm) < 0)) {
+	if ((!pcm->draining) && (!pcm->running) && (pcm_start(pcm) < 0)) {
 		return -errno;
-	} else {
-		pcm->running = 1;
 	}
 
-	/* Wait for deque message from kernel */
-	size = mq_receive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio);
-	if (size != sizeof(msg)) {
-		/* Interrupted by a signal? What to do? */
-		return oops(pcm, EINTR, "Interrupted while waiting for deque message from kernel");
-	}
-	if (msg.msgId == AUDIO_MSG_DEQUEUE) {
-		apb = (struct ap_buffer_s *)msg.u.pPtr;
-		/* Copy data to user buffer */
-		memcpy(data, apb->samp, apb->nbytes);
-		/* Enque buffer for next read opertion */
-		nbytes = apb->nbytes;
-		apb->nbytes = 0;
-		apb->curbyte = 0;
-		apb->flags = 0;
-		bufdesc.u.pBuffer = apb;
-		if (ioctl(pcm->fd, AUDIOIOC_ENQUEUEBUFFER, (unsigned long)&bufdesc) < 0) {
-			return oops(pcm, errno, "failed to enque buffer after read");
+	/* Read data */
+	if (pcm->next_buf != NULL) {
+		/* We have a pending buffer with some data in it. */
+		apb = pcm->next_buf;
+
+		if (nbytes < pcm->next_size) {
+			/* User wants less data than we have in buffer.
+			   Copy the requested amount of data to user buffer and return */
+			memcpy(data, apb->samp + pcm->next_offset, nbytes);
+			pcm->next_size -= nbytes;
+			pcm->next_offset += nbytes;
+		} else {
+			/* User wants more data than we have in buffer.
+			   Copy available data to user buffer, enqueue buffer and return */
+			memcpy(data, apb->samp + pcm->next_offset, pcm->next_size);
+
+			nbytes = pcm->next_size;
+			pcm->next_size = 0;
+			pcm->next_offset = 0;
+			pcm->next_buf = NULL;
+
+			apb->nbytes = 0;
+			apb->curbyte = 0;
+			apb->flags = 0;
+			bufdesc.u.pBuffer = apb;
+			if (ioctl(pcm->fd, AUDIOIOC_ENQUEUEBUFFER, (unsigned long)&bufdesc) < 0) {
+				return oops(pcm, errno, "failed to enque buffer after read");
+			}
 		}
+
 	} else {
-		return oops(pcm, EINTR, "Recieved unexpected msg (id = %d) while waiting for deque message from kernel", msg.msgId);
+		/* We got here because we dont have any pending buffers */
+		/* Wait for deque message from kernel */
+		if (pcm->draining) {
+			/* When we are in draining state, we will check for buffers already in the msg queue
+			   If there is no buffer on the queue, we return immediately */
+			clock_gettime(CLOCK_REALTIME, &st_time);
+			size = mq_timedreceive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio, &st_time);
```

Comment:
Can we miss any message here due to some scheduling or latency issues?
Wouldn't it be better to maintain the number of buffers enqueued and then deque them in each pcm_readi while draining

Reply:
We wont miss any buffers due to timing issues, because once a message is put on the mq, it will remain there. Also, we cannot keep count of enqueued buffers. We may enqueue 4 buffers and then call drain. By the time we call drain, the codec has filled only 2 buffers. When drain is called, codec is immediately stopped. So, Only 2 buffers can be dequeued. Since we wont know how many buffers will be filled by codec when we call drain(), we cannot keep count.

Reply:
What I meant here was if the pcm_readi calls are made quickly then the next message may not arrive at mq_timedreceive.
This message may not arrive due to large buffer size or due to latency/scheduling problems.  So we may miss the message and return oops.

What I thought of drain was to get audio data from all the enqueued buffers and to stop enqueuing buffers for further data and call pcm_stop are getting all the data. If that is not what drain means , then the in flight thing will not work.

Reply:
Since this is in readi part, we would have already stopped the recording before we get here. After this point the driver will not send any more messages on the queue. This mq receive is done only to empty out any messages that are already present. 
However, after this point we dont need to enqueue the buffers again. That point is correct. But I just decided to keep the code simple. Even if we call enqueue, it does not matter since the codec has already been stopped and we have set our local variables like pcm->running and pcm->but_ptr to zero.